File: src/main/java/jukebot/commands/PlayerStats.java
package jukebot.commands;

import jukebot.audio.AudioHandler;
import jukebot.utils.Command;
import jukebot.utils.CommandProperties;
import jukebot.utils.Context;

import java.text.DecimalFormat;

@CommandProperties(description = "Displays player statistics", aliases = {"ps"}, developerOnly = true)
public class PlayerStats implements Command {

    private DecimalFormat dpFormatter = new DecimalFormat("0.00");

    @Override
    public void execute(Context context) {
        AudioHandler player = context.getAudioPlayer();

        double packetLossPc = (double) player.trackPacketLoss / player.trackPackets * 100;

        context.getChannel().sendMessage("Dropped packets: " + player.trackPacketLoss +
                "\nSent packets: " + player.trackPackets +
                "\n% Lost: " + dpFormatter.format(packetLossPc)).queue();
    }

}

File: src/main/java/jukebot/commands/PlayerStats.kt
package jukebot.commands

import jukebot.audio.AudioHandler
import jukebot.framework.Command
import jukebot.framework.CommandCheck
import jukebot.framework.CommandProperties
import jukebot.framework.Context
import java.text.DecimalFormat

@CommandProperties(description = "Displays player statistics", aliases = ["ps"], developerOnly = true)
@CommandCheck(isPlaying = true)
class PlayerStats : Command(ExecutionType.REQUIRE_MUTUAL) {

    private val dpFormatter = DecimalFormat("0.00")

    override fun execute(context: Context) {
        val player = context.getAudioPlayer()

        val packetsLost = player.trackPacketLost.toDouble() / player.trackPacketsSent * 100
        val packetLossPc = dpFormatter.format(packetsLost)

        val trackProgress = player.player.playingTrack.position.toDouble() / 60000
        val totalPackets = player.trackPacketsSent + player.trackPacketLost
        val expectedPackets = AudioHandler.EXPECTED_PACKET_COUNT_PER_MIN * trackProgress
        val framesDeficit = Math.abs(expectedPackets - totalPackets).toInt()

        context.channel.sendMessage(
            "Dropped packets: ${player.trackPacketLost} ($packetLossPc%)\n" +
                "Deficit packets: $framesDeficit\n" +
                "Sent packets: ${player.trackPacketsSent}"
        ).queue()
    }

}
